cloudinit/sources/DataSourceOpenNebula.py

```diff
@@ -356,29 +448,18 @@ def varprinter(vlist):
         + varprinter(allvars)
         + "{\n%s\n\n:\n} > /dev/null\n" % content
         + "unset IFS\n"
-        + varprinter(keylist)
+        + varprinter(allvars)
```

Comment:
> Does this version even work? 

I think so. I had to tear it apart to rebuild it in posix shell, and I recall that this did what I think it was supposed to. I think that the line containing: `for line in output.split("\x00"):` is crucial to null-delimited parsing stuff.

> Is there a reason we can't replace the whole varprinter mess with set -o posix; set?

I was trying to avoid disturbing sleeping dragons with excessive changes, but I just checked[1] and bash's `set` output multi-line strings rather than ansi-c strings, so it probably would work. We'd just need to unset that `-o posix` if it wasn't already set. Bash stores shell options in `SHELLOPTS`, so I think we have all of the bits required to make this a single codepath.

I previously thought that ansi-c multi-line strings was technically posix-compliant[2] and possible output from `set` as "appropriate quoting". Posix shells like dash and ash certainly understand them. However, digging further it seems that maybe it's [actually an extension that was backported](https://unix.stackexchange.com/a/371873/219645) into these shells for support, which is why `-o posix` disables this output format.

[1] bash with posix -o
```
bash -c "TEST=$'multi-\nline'; set" | grep TEST
BASH_EXECUTION_STRING='TEST=$'\''multi-\nline'\''; set'
TEST=$'multi-\nline'
bash -c "TEST=$'multi-\nline'; set -o posix; set" | grep TEST
BASH_EXECUTION_STRING='TEST=$'\''multi-\nline'\''; set -o posix; set'
```

[2] [the spec](https://pubs.opengroup.org/onlinepubs/9699919799/utilities/V3_chap02.html#set)
> If no options or arguments are specified, [set](https://pubs.opengroup.org/onlinepubs/9699919799/utilities/V3_chap02.html#set) shall write the names and values of all shell variables in the collation sequence of the current locale. Each name shall start on a separate line, using the format:
> 
> "%s=%s\n", <name>, <value>
>
> The value string shall be written with appropriate quoting; see the description of shell quoting in [Quoting](https://pubs.opengroup.org/onlinepubs/9699919799/utilities/V3_chap02.html#tag_18_02). The output shall be suitable for reinput to the shell, setting or resetting, as far as possible, the variables that are currently set; read-only variables cannot be reset.

Reply:
> Is there a reason we can't replace the whole varprinter mess with set -o posix; set?

It looks like that works.